reconciler/reconciler.go

```diff
@@ -184,3 +157,76 @@ func (r *reconciler[Obj]) loop(ctx context.Context, health cell.Health) error {
 		}
 	}
 }
+
+// prune performs the Prune operation to delete unexpected objects in the target system.
+func (r *reconciler[Obj]) prune(ctx context.Context, txn statedb.ReadTxn) error {
+	iter, _ := r.Config.Table.All(txn)
+	start := time.Now()
+	err := r.Config.Operations.Prune(ctx, txn, iter)
+	if err != nil {
+		err = fmt.Errorf("prune failed: %w", err)
+	}
+	r.metrics.PruneDuration(r.ModuleID, time.Since(start))
+	r.metrics.PruneError(r.ModuleID, err)
+
+	return err
+}
+
+func (r *reconciler[Obj]) refreshLoop(ctx context.Context, _ cell.Health) error {
+	refreshTicker := time.NewTicker(r.Config.RefreshInterval)
+	defer refreshTicker.Stop()
+
+	for {
+		// Wait until it's time to refresh.
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-refreshTicker.C:
+		}
+
+		// Collect all objects that need a refresh, e.g. have not been refreshed in RefreshInterval
+		// or longer. We collect them as queries to avoid holding onto potentially many objects for
+		// a long period of time.
+		type entry struct {
+			q   statedb.Query[Obj]
+			rev statedb.Revision
+		}
+		toRefresh := []entry{}
+		iter, _ := r.Config.Table.All(r.DB.ReadTxn())
+		indexer := r.Config.Table.PrimaryIndexer()
+		now := time.Now()
+		for obj, rev, ok := iter.Next(); ok; obj, rev, ok = iter.Next() {
+			status := r.Config.GetObjectStatus(obj)
+			if status.Kind == StatusKindDone && now.Sub(status.UpdatedAt) >= r.Config.RefreshInterval {
+				toRefresh = append(toRefresh, entry{indexer.QueryFromObject(obj), rev})
+			}
+		}
+
+		if len(toRefresh) == 0 {
+			continue
+		}
+
+		// Evenly spread the refreshes over a period of RefreshInterval/2 to avoid
+		// disruptions to normal reconciliation.
+		markTicker := time.NewTicker(r.Config.RefreshInterval / 2 / time.Duration(len(toRefresh)))
```

Comment:
We discussed this offline, what `RefreshInterval` really means (i. e. what is the expected time for actual refresh / worst case time), if we want to have  `RefreshInterval / 2` in this line or maybe just `RefreshInterval` etc.

How about a totally different design, a little bit less discrete compared to this one:
- we have one "consumer" that pulls requests from workqueue to refresh with expected qps of ~`Table Elements / RefreshInterval`, but checks if the object needs updating (if it wasn't update in mean-time) before marking it as pending.
- we have one "producer" that watches for Table changes and adds them to something like delaying workqueue: https://github.com/kubernetes/client-go/blob/master/util/workqueue/delaying_queue.go

This would mean that expected refresh time is essentially almost always smaller than `~RefreshInterval`, it would evenly spread updates over time, while also ensuring that work is not wasted in case the object gets updated in mean-time.

Reply:
That's somewhat similar to what I initially had in mind, e.g. have a priority queue key'd on when the object should refresh, but I ended up going for the simpler design. I can take a look at how this would look like. I don't want to use the client-go workqueue as there's users of StateDB that don't want to pull in all those extra dependencies, but I can reuse the `retryQueue` priority queue implementation for this.